### ISSUE TITLE
The iterator need getList to return a lazy loaded collection

### ISSUE THREAD

**Issue by peterlembke** [magento/magento2#15155]
The standard repository function getList do not always use lazy loading of a collection.
When you use an iterator on a collection to reduce the memory usage, then it works really well only on those collections that use lazy loading.

### Preconditions
All Magento versions including 2.3

### Steps to reproduce
By using
\Magento\Framework\Model\ResourceModel\Iterator
with getList() from any of:
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Sales/Model/Order/AddressRepository.php#L95
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Sales/Model/Order/InvoiceRepository.php#L98
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Sales/Model/Order/CreditmemoRepository.php#L103
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Sales/Model/Order/ShipmentRepository.php#L94

then it works really well because getList uses lazy loading of the collection.

$this->iterator->walk($this->_collection->getSelect(), [[$this, 'executeFullItem']], ['updatedIndexerIds' => &$updatedIndexerIds]);

Other repositories have a different solution:
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Customer/Model/ResourceModel/CustomerRepository.php#L305
https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Customer/Model/ResourceModel/AddressRepository.php#L177

They load the collection, make an array of them and attach the array to the result.
As soon as the collection is used in a for-loop, all the data are loaded into memory. No more lazy loading.

This approach makes it impossible to use the iterator to save memory.

### Expected result
Expect getList to return a collection that is not loaded yet.
One could expect that resource collections should work the same everywhere.

### Actual result
On some repository getList you get a collection that is not loaded. (Expected)
On some repository getList you get an array with items.

### Lazy loading
There is a function ->getItems() that is expected to load the collection.
You can also use count($collection) or use the collection in a for-loop, both will load the collection into memory.
The iterator is designed to load one item at the time from a lazy loaded (not yet loaded) collection. Then you can manipulate that item and load the next one. This can have a huge impact in reducing memory consumption.

### Suggested solution
For CustomerRepository, and for Customer/AddressRepository,
public function getList(SearchCriteriaInterface $searchCriteria, $legacy = true)
after
$collection->setPageSize($searchCriteria->getPageSize());

if ($legacy === false) {
	return $collection;
}

That will not break any code and it is still possible to get the lazy collection.

**Comment by ghost** [comment]
@peterlembke,  thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by bartoszkubicki** [comment]
@magento-engcom-team 
Should be $collection really returned? Maybe SearchResult in general could implement interface `Iterator` or `IteratorAggregate`?

**Comment by peterlembke** [comment]
The damage is already done as soon as the collection are loaded into memory by the for-loop inside of getList().
It does not matter what getList() returns as long as all getList() return the same type of object.
That can not be achieved since each repository declare their own getList() and can return whatever they want.
A repository developer can omit implementing a getList() function, or name the function whatever they want.
This battle can not be won as long as Magento 2 are structured this way.
I guess we will have to live with this quirk too and extend the project budget, let the customer pay for the extra development time and the extra hardware needed.

As a parallel, Magento 1 was structured differently, there all collections could be iterated.

If there is a general solution, then me and my customers would be very grateful.

**Comment by bartoszkubicki** [comment]
You are right, maybe this could be the first step - refactoring this two classes. Loops should be removed, as they even can exhaust memory, by calling getList().
In my opinion we should have some `AbstractRepository`, which would serve some basic function like `get`, `getById`, `save`, `delete`, `deleteById`. Then while creating repository with standard logic, we would not have to implement tons of boilerplate code for repositories - we have some abstract SearchResults. And if you need something else you inherit from it and override it or doesn't inherit at all and only implement your interface and do things in your way.

**Comment by peterlembke** [comment]
An AbstractRepository would be awesome. Yes, refactoring the two functions would be a good start.

**Comment by magento-engcom-team** [comment]
Hi @milindsingh. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by peterlembke** [comment]
Good luck M2 community. I switched to Laravel last year after 8y with Magento.

**Comment by tuyennn** [comment]
@magento-engcom-team @sdzhepa What is the current status of this 3year-olds suggestion improvement on current `2.4-develop` ?

**Comment by bartoszkubicki** [comment]
Still a possible improvement
